### DISCUSSION TITLE
macOS install problem - `AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'`

### DISCUSSION THREAD

**Discussion by tedthetrumpet** [DigiScore/neoscore#35] [Q&A] [answered]
I am hitting some sort of problem with the install on macOS 12.4, as below. Any ideas?

```
Last login: Wed May 18 10:29:56 on console
jsimon@jsairm ~ % pip install neoscore
Collecting neoscore
  Downloading neoscore-0.1.0-py3-none-any.whl (932 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 932.8/932.8 kB 7.2 MB/s eta 0:00:00
Collecting sortedcontainers==2.4.0
  Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)
Collecting PyQt5<6.0.0,>=5.15.6
  Downloading PyQt5-5.15.6.tar.gz (3.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 9.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [29 lines of output]
      Traceback (most recent call last):
        File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 156, in prepare_metadata_for_build_wheel
          hook = backend.prepare_metadata_for_build_wheel
      AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
          main()
        File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 160, in prepare_metadata_for_build_wheel
          whl_basename = backend.build_wheel(metadata_directory, config_settings)
        File "/private/var/folders/lg/3c1b35bj7v7685m629_tfqkh0000gn/T/pip-build-env-q7i_oaoj/overlay/lib/python3.10/site-packages/sipbuild/api.py", line 51, in build_wheel
          project = AbstractProject.bootstrap('pep517')
        File "/private/var/folders/lg/3c1b35bj7v7685m629_tfqkh0000gn/T/pip-build-env-q7i_oaoj/overlay/lib/python3.10/site-packages/sipbuild/abstract_project.py", line 83, in bootstrap
          project.setup(pyproject, tool, tool_description)
        File "/private/var/folders/lg/3c1b35bj7v7685m629_tfqkh0000gn/T/pip-build-env-q7i_oaoj/overlay/lib/python3.10/site-packages/sipbuild/project.py", line 594, in setup
          self.apply_user_defaults(tool)
        File "/private/var/folders/lg/3c1b35bj7v7685m629_tfqkh0000gn/T/pip-install-eakcgivf/pyqt5_746201daaf88444cb717f31562acb8d1/project.py", line 63, in apply_user_defaults
          super().apply_user_defaults(tool)
        File "/private/var/folders/lg/3c1b35bj7v7685m629_tfqkh0000gn/T/pip-build-env-q7i_oaoj/overlay/lib/python3.10/site-packages/pyqtbuild/project.py", line 70, in apply_user_defaults
          super().apply_user_defaults(tool)
        File "/private/var/folders/lg/3c1b35bj7v7685m629_tfqkh0000gn/T/pip-build-env-q7i_oaoj/overlay/lib/python3.10/site-packages/sipbuild/project.py", line 241, in apply_user_defaults
          self.builder.apply_user_defaults(tool)
        File "/private/var/folders/lg/3c1b35bj7v7685m629_tfqkh0000gn/T/pip-build-env-q7i_oaoj/overlay/lib/python3.10/site-packages/pyqtbuild/builder.py", line 67, in apply_user_defaults
          raise PyProjectOptionException('qmake',
      sipbuild.pyproject.PyProjectOptionException
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
jsimon@jsairm ~ % python3 --version
Python 3.10.4
jsimon@jsairm ~ % 


```

**Comment by ajyoon** [comment]
Looks like this is [a known issue with PyQt and M1 chips](https://stackoverflow.com/questions/65901162/how-can-i-run-pyqt5-on-my-mac-with-m1chip). Could you try running `pip install --upgrade pip` and then try installing neoscore again?

**Reply by ajyoon** [reply]
assuming this is the fix, [I've updated the troubleshooting guide accordingly](https://neoscore.org/community/support.html).

**Reply by tedthetrumpet** [reply]
No good, still same error. Looking at that page you linked to on stackoverflow, the position with M1 chips seems a bit iffy at the moment. I have an idea: I have a machine running UbuntuStudio, maybe the install would be ok on that instead? Worth a try, do you think?

**Reply by tedthetrumpet** [reply]
```
jsimon@jsairm ~ % pip install --upgrade pip
Requirement already satisfied: pip in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (22.1)
 
```

**Reply by ajyoon** [reply]
Ubuntu should work fine, but if you're able I would like to figure this issue out so we can add it to the troubleshooting guide. Does this help?

```
pip install --upgrade setuptools
```

**Reply by ajyoon** [reply]
[it seems like lots of people have this error from all kinds of environments](https://duckduckgo.com/?t=ffab&q=AttributeError%3A+module+%27sipbuild.api%27+has+no+attribute+%27prepare_metadata_for_build_wheel%27&ia=web), some strange upstream PyQt issue that keep cropping up in different PyQt5+environment combinations. It even happens on PyQt6 sometimes.

**Reply by tedthetrumpet** [reply]
```
jsimon@jsairm ~ % pip install --upgrade setuptools
Requirement already satisfied: setuptools in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (58.1.0)
Collecting setuptools
  Using cached setuptools-62.3.1-py3-none-any.whl (1.2 MB)
Installing collected packages: setuptools
  Attempting uninstall: setuptools
    Found existing installation: setuptools 58.1.0
    Uninstalling setuptools-58.1.0:
      Successfully uninstalled setuptools-58.1.0
Successfully installed setuptools-62.3.1

```
but after that `pip install neoscore  ` still gives same error.

**Reply by ajyoon** [reply]
hmm.. could you please confirm which pip version you're installing from? `pip --version` - sometimes `pip` can point to the wrong python environment. Does `python3 -m pip --version` give any different result?

**Reply by ajyoon** [reply]
also, if you're not running in a [virtual environment](https://realpython.com/python-virtual-environments-a-primer/) could you try setting one up and installing from within that? package installation on the global python interpreter is not recommended since it can cause all kinds of weird conflicts with system-wide package.

**Reply by ajyoon** [reply]
and just another thought, you get the same error when installing PyQt5 directly, right? `pip install PyQt5==5.15.6` - or if that doesn't work, what about just `pip install PyQt5`?

**Reply by tedthetrumpet** [reply]
At this point I should explain that I'm a musician not a programmer! So some of this is a little unfamiliar. Following the instructions at https://realpython.com/python-virtual-environments-a-primer/ I've done `python3 -m venv venv`. I don't understand the next bit about 'activate the environment by executing a script that comes with the installation' and 'make sure that you’re in the folder that contains the virtual environment you just created'. What installation? What folder should I be in?

**Reply by tedthetrumpet** [reply]
Ok, I do now seem to have a directory `/Users/jsimon/venv/` that looks to me like it has a Python 10 install in it. So far so good. And I can also see `/Users/jsimon/venv/bin/activate`. What do I do now?

**Reply by tedthetrumpet** [reply]
Also, sorry, missed the question about the pip version. Looks ok to me:

```
jsimon@jsairm ~ % pip --version
pip 22.1 from /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip (python 3.10)

```

**Reply by ajyoon** [reply]
Ok, sounds like your environment has been provisioned, you just need to activate it. This can be done with `source venv/bin/activate` (ensuring the path matches that of the activate file). When activated, try installing neoscore again from there.

**Reply by tedthetrumpet** [reply]
Ok, I think this looks right:

```
jsimon@jsairm venv % source /Users/jsimon/venv/bin/activate 
(venv) jsimon@jsairm venv % 
```

**Reply by tedthetrumpet** [reply]
Ok, in that environment it is telling me pip is out of date, let me try to update pip there… done that… `Successfully installed pip-22.1` try again… `pip install neoscore` sorry, still that same error :(

**Reply by ajyoon** [reply]
Very weird. Are you able to install PyQt5 directly? `pip install PyQt5`

**Reply by tedthetrumpet** [reply]
Nope, sorry, tried that it gives exactly that same error about `AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'`

**Reply by ajyoon** [reply]
hmmm.. [could you try following the instructions here to run from within Rosetta](https://stackoverflow.com/a/70262165/5615927)? sorry this is such a hassle, it's an upstream issue with PyQt on the latest Mac architecture many people have run into

**Reply by tedthetrumpet** [reply]
Well, that's a slightly terrifying series of instructions, but so far so good: `Successfully installed PyQt5-5.15.6 PyQt5-Qt5-5.15.2 PyQt5-sip-12.10.1`

**Reply by tedthetrumpet** [reply]
Looks good!

**Reply by ajyoon** [reply]
Excellent! I'll update the troubleshooting guide to include this case. Thanks for sticking with it, and I hope you have fun playing around with the library!

**Reply by tedthetrumpet** [reply]
Yup, thanks, the 'hello world' program at least seems fine :)

**Comment by ajyoon** [comment, marked as answer]
[For future reference, the solution to this problem is now documented here](https://neoscore.org/community/support.html#pyqt5-installation-problems).

**Comment by richardhoadley** [comment]
Hi, I'm trying to install neoscore on a Raspberry Pi running Raspbian v10 (buster).  I'm getting exactly the same error(s) as above and I've tried most of the above suggestions, except, of course, that I have no Rosetta to run from!  Is there anything that I can do to progress?  I have, incidentally, managed to get neoscore working on a Mac running 10.13.6.

**Reply by craigvear** [reply]
Hi Richard. PyQT is a dependency for Neoscore, but getting it to work on Raspian needs some additional work beyond the Pip install. I have followed this previously with some success https://raspberrypi.stackexchange.com/questions/62939/pyqt5-on-a-raspberry-pi#63058

I had to install pyqt5 first, checking it, then install Neoscore.

**Comment by ajyoon** [comment]
FYI all, (@tedthetrumpet @richardhoadley) we just released neoscore 0.1.3, which reduces the minimum required Python version to 3.7 and should improve PyQt support since there are prebuilt wheels for 3.7. M1 chips still require Rosetta, but otherwise we broadly expect this should fix most PyQt-related installation issues.

**Comment by richardhoadley** [comment]
Hi Craig

Thanks for responding so quickly!  I’ve had no luck with this so far, but will have time to continue in a few days and will post again if I have any success.  Just a quick question in the meantime, should I install pyqt5 in the main programme space, or the virtual space I’ve created for neoscore?  Sorry if this is a silly question, but I’m not used to serious programming in Python – the most I’ve done is helping my daughter (12 years old at the time!) write a (text only) game in Python for her home work, and that was a couple of years ago…

Best

Richard

**Reply by ajyoon** [reply]
I recommend confirming you're installing pyqt / neoscore on python 3.7 by running `python --version` before attempting to `pip install`. In general we recommend following standard Python best practices by making a dedicated virtual environment for each project and installing neoscore and other dependencies in that environment. [There is a good guide for using virtual environments here](https://realpython.com/python-virtual-environments-a-primer/). That said, because Raspberry Pi's are often used for a single project each, it may not be such a big deal to cheat with it.
